[2단계 - 로또(수동)] 야호(최지연) 미션 제출합니다.

yaho99 opened the pull request:
안녕하세요 희봉! 2단계로 돌아온 야호입니다 :)
1단계 마지막 리뷰에서 주신 내용도 수정하고, 수동 로또도 구현했어요!

미션을 진행하며 궁금했던 점들은 코멘트로 달아두겠습니다 ㅎㅎ
이번 피드백도 잘 부탁드려요~!~!

+ 머지된 지난 코드에서 LottoTest에서 주신 질문에 대해 코멘트 달아두었습니다 :)

yaho99 committed:
docs: 희봉의 두번째 리뷰에 따른 수정사항 정리

yaho99 committed:
refactor: Number, WinningNumber 에 대한 equals() 를 오버라이딩해 같은 숫자일 경우 같은 값으로 인지하도록 리팩터링

yaho99 committed:
refactor: Number, WinningNumber 에 대한 equals() 를 오버라이딩해 같은 숫자일 경우 같은 값으로 인지하도록 리팩터링

yaho99 committed:
refactor: getPrize()와 getPrizeByCount() 를 하나로 통합

yaho99 committed:
docs: 2단계 미션을 위한 기능 요구사항 작성

yaho99 committed:
style: 로또에 사용되는 숫자를 Lotto Ball 로 변경

yaho99 committed:
style: 기존의 Lotto 를 AutoLotto 로 변경

yaho99 committed:
refactor: 상수에 private static final 추가

yaho99 committed:
refactor: 패키지 이동 후 불필요한 import 문 삭제

yaho99 committed:
feat: AutoLotto 와 ManualLotto 의 기반이 될 Lotto 생성, 예외 처리

1. 6개의 LottoBall 로 구성
2. 서로 다른 LottoBall 로 구성

yaho99 committed:
feat: LottoBall 을 정렬하기 위해 compareTo 오버라이드

yaho99 committed:
refactor: Lotto 의 정적 팩터리 메서드를 생성하고, 생성자를 private 으로 변경

yaho99 committed:
refactor: AutoLotto 에서 로또의 숫자 타입으로 사용되던 Integer 를 LottoBall 로 변경

yaho99 committed:
refactor: 입력 값에 포함된 공백 처리

yaho99 committed:
refactor: AutoLotto 에서 로또의 숫자 타입으로 사용되던 Integer 를 LottoBall 로 변경

yaho99 committed:
refactor: AutoLotto 를 Lotto 로 변경하고, 수동 생성 로또 기능 추가

yaho99 committed:
refactor: 로또 개수 입력에 대한 유효 검사를 InputView 에서 진행하여 예외 입력시 바로 에러 처리

yaho99 committed:
refactor: 기존에 생성해둔 private 메서드 활용

yaho99 committed:
refactor: AutoLotto 를 Lotto 로 변경

yaho99 committed:
style: Number 를 LottoBall 로 변경

yaho99 committed:
style: Number 를 LottoBall 로 변경

yaho99 committed:
refactor: 로또 구매 내역 출력

yaho99 committed:
style: 오름차순으로 재정렬

yaho99 committed:
refactor: 로또 개수 입력에 대한 유효 검사를 InputView 에서 진행하며 불필요해진 테스트 삭제

yaho99 committed:
refactor: LottoBall 이 생기며 불필요해진 WinningBall 삭제, LottoBall 로 대체

yaho99 committed:
refactor: 예외 입력시 에러 메세지를 출력하고, 재입력을 요구

yaho99 committed:
style: 입력 요구 사항 간 간격 추가

yaho99 committed:
docs: 2단계 미션을 위한 요구사항 목록 체크

yaho99 committed:
style: import 문 정리, 문맥에 맞는 공백 추가, 사용 순서에 따라 상수 정리

yaho99 committed:
style: 의미를 명확하게 나타내는 변수명으로 변경

yaho99 committed:
style: 의미를 명확하게 나타내는 패키지명으로 변경

yaho99 committed:
docs: github 소스코드 탭사이즈 설정

heebong commented on src/test/java/lotto/model/LottoTest.java:
```diff
@@ -0,0 +1,103 @@
+		);
+		Lotto lotto2 = new Lotto(lottoBalls2);
+
+		assertThat(lotto1).isNotEqualTo(lotto2);
```
`Lotto` 객체에 equals, hashcode 함수를 오버라이딩하면 같은 객체로 인식할겁니다. (이미 `LottoBall`은 해당 함수가 오버라이딩 되어있으니 `Lotto`만 하면 돼요)
물론 야호가 판단하기에 다른 로또라고 생각되면 그렇게 개발하면 됩니다 😄

heebong commented on src/main/java/lotto/model/ball/WinningBalls.java:
```diff
@@ -0,0 +1,51 @@
+	private static final String ERROR_DUPLICATE = "[ERROR] 당첨 번호는 중복되면 안됩니다";
+	private static final int WINNING_NUMBERS_SIZE = 6;
+
+	private final List<LottoBall> winningBalls;
```
`WinningBalls`와 `Lotto`객체의 차이는 무엇일까요?
둘 다 `List<LottoBall>`의 일급컬렉션입니다. 같은 일급컬렉션이 두개가 있어야할까요?

yaho99 replied:
그러게요... 😂 전혀 생각을 못했어요...!

그래도 `WinningBalls` 라는 객체는 `Lotto` 와 다르게 우승 번호라는 역할을 가지고 따로 있었으면 좋겠는데,
이런 경우 `WinningBallls` 를 `List<LottoBall> winningBalls` 가 아닌 `Lotto winningBalls` 로 만들면 괜찮을까요??

heebong commented on src/test/java/lotto/model/MoneyTest.java:
```diff
@@ -1,9 +1,6 @@
+import static org.assertj.core.api.Assertions.*;
 
 import org.junit.jupiter.api.DisplayName;
 import org.junit.jupiter.api.Test;
```
`Money#rate` 함수의 테스트가 없네요 👀

heebong commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -16,19 +16,34 @@ private Lottos(List<Lotto> lottos) {
 	}
 
-	public static Lottos purchase(int count) {
+	public static Lottos purchase(int totalCount, int manualCount, List<String[]> manualLottoNumbers) {
```
`purchase`는 테스트코드가 없네요. 너무 많은 걸 하고 있어서 테스트하기 힘든 함수가 된 것 같은데 나눠볼 수 있을까요?
1. 함수로 나누기
2. 클래스로 나누기

yaho99 replied:
제가 autoCount 를 구하는 메서드를 분리해보고, 테스트를 진행해보기는 했는데 희봉이 원하시는 내용인지는 잘 모르겠어요... :(

함수로 나눈다는 것은 이해가 되는데, 클래스로 나눈다는 것은 무슨 뜻인지 잘 모르겠습니다!
purchase 라는 클래스를 따로 작성하라는 말씀이신가요?!

제가 보기에는 purchase 가 그렇게 많은 일을 하고 있다고 생각되지 않는데, 
희봉이 보시기에는 어떤 부분이 많은 일을 하고 있다고 말씀하신건지 궁금합니다 :)

heebong commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -16,19 +16,34 @@ private Lottos(List<Lotto> lottos) {
 	}
 
-	public static Lottos purchase(int count) {
+	public static Lottos purchase(int totalCount, int manualCount, List<String[]> manualLottoNumbers) {
```
추가로 배열말고 리스트를 사용하자!

heebong commented on src/main/java/lotto/model/prize/Prize.java:
```diff
@@ -21,18 +21,8 @@
-	private static Prize getPrizeByCount(MatchResult matchResult) {
 		return Arrays.stream(values())
-				.filter(prize -> matchResult.isCount(prize.matchCount))
+				.filter(prize -> matchResult.isCount(prize.matchCount) && matchResult.isBonus() == prize.bonus)
```
`1,2,3,4,5,6` 로또를 샀지만 당첨번호 `1,2,3,11,12,13` 보너스번호 `4`일 경우 5등을 못할 것 같네요 😢

heebong reviewed (changes requested):
안녕하세요, 야호! 돌아온 리뷰어 희봉입니다 😎 
2단계 미션을 잘 구현하셨네요! 요구사항이 변경되어 힘들었을텐데 잘 개발하셨습니다 👍 
몇가지 생각할거리를 코멘트로 남겨놓았어요. 확인 부탁드립니다. 궁금한게 있을 경우 댓글 혹은 DM 아시죠? 😄

yaho99 committed:
test: Stream 중간 연산자, 종단 연산자에 대한 학습 테스트

yaho99 committed:
test: String 과 StringBuilder 에 대한 학습 테스트

yaho99 committed:
refactor: stream 으로 대체 가능한 for 문 수정

yaho99 committed:
test: Money 의 rate() 에 대한 테스트 추가

yaho99 committed:
refactor: 배열 대신 List 사용

yaho99 committed:
refactor: WinningBalls 를 Lotto winningBalls 로 변경

yaho99 committed:
refactor: 관련된 클래스끼리 다시 패키징

yaho99 committed:
refactor: 자동 로또와 수동 로또 생성자를 각각 정적 팩터리 메서드로 생성

yaho99 committed:
test: Lotto 의 countTickets(), match(), contains() 에 대한 테스트 진행

yaho99 committed:
fix: 2등이 아니어도 보너스가 일치하는 경우 2등이 반환되는 오류 수정

yaho99 committed:
test: Lotto 의 purchase() 가 입력된 만큼 로또를 생성하는지 테스트

yaho99 committed:
refactor: 메서드의 역할을 줄이기 위한 메서드 분리

yaho99 committed:
refactor: prizeInformation 을 EnumMap 으로 변경

yaho99 committed:
test: BonusBall 의 생성자, match() 에 대한 테스트 추가

yaho99 committed:
refactor: 특정 범위에 대한 for 문을 IntStream 으로 변경

yaho99 committed:
style: 자동 정렬, 임포트 정리 기능 사용

yaho99 committed:
style: 자동 정렬, 임포트 정리 기능 사용

yaho99 committed:
refactor: 메서드 설명 중 오타 수정

yaho99 committed:
docs: 희봉의 세번째 리뷰에 따른 수정사항, 개인적 리팩터링 사항 정리 및 체크

yaho99 commented:
안녕하세요 희봉! 좋은 리뷰들 항상 감사드립니다 :)
항상 더 고칠게 있을까 싶은데 희봉의 리뷰를 보면 아차 싶어서 놀라워요...ㅎㅎ
정말 매의 눈을 가지신 것 같아요  👀

이번에는 희봉의 리뷰들도 적용해보고, 개인적으로 리팩터링 해보고 싶은 것들도 추가해서 진행해 보았어요!!
희봉이 보시기에 마음에 안들게 퇴화 되었을 수도 있지만... 하핫.... 😂
그렇다면 마구 지적 해주세요 :) 열심히 배우겠습니다!!

희봉이 보시기 편하게 제가 적용해보고 싶었던 것들을 정리해 볼게요!

<hr>

## EnumMap 을 적용해 보았어요!

상금 정보에는 등수(prize)와 해당 등수의 개수가 들어가는데, 이걸 map 을 이용해보면 어떨까 생각하던 중, EnumMap 에 대해 알게 되었어요!
제가 이펙티브 자바 스터디를 하는 중인데 마침 한 크루가 해당 아이템을 설명해주셔서 다른 크루들의 코드를 찾아보니,
이미 EnumMap 을 사용한 크루들도 있더라구요!

그래서 저도 한번 적용해  봤는데, 잘 사용한건지 모르겠네요...ㅎ

그래도 불필요한 클래스 하나가 줄었고, 그만큼 불필요한 연산이 줄었다고 생각해서 저는 만족스러웠습니다!

그리고 진행해보니 생각보다 큰 리팩터링이 되었는데, DTO 를 만들어둬서 그런지 Controller 나 View 에는 변화가 전혀 없었어요!
DTO 의 필요성을 깨닫게 되는 계기도 되어서 좋았습니다. ㅎㅎ

희봉은 이 변화를 어떻게 생각하시는지, 수정해야 할 부분이 있어 보이시는지 궁금합니다 :)

## 자동 로또와 수동 로또에 대한 정적 팩터리 메서드를 따로 만들어 봤어요!

정적 팩터리 메서드를 사용하는 이유를 공부하던 중, 이름을 가질 수 있다는 내용을 보게 되었어요

'이름을 가질 수 있다는게 장점이라면, 생성 방식이 다른 두가지 메서드를 가질 수 있다는 것도 장점이 아닐까?' 라는 생각에
자동 로또와 수동 로또에 대한 정적 팩터리 메서드를 각각 생성해 보았어요!

생성을 하고 나서 에러가 뜬 테스트 코드들에 대한 리팩터링도 진행했는데, 테스트 코드도 훨씬 간단해져서 좋았습니다.
그리고 해당 메서드를 사용하는 곳에서도 비교적 명시적으로 역할을 알 수 있게되어 가독성도 높아진 것 같습니다.

저는 이런 장점들을 경험했는데, 제가 놓치고 있는 단점이 있다면 알려주시면 감사하겠습니다 :)

## for 문을 최대한 stream 으로 변경해 작성해 보았어요!

이번 미션을 시작하기 전, 제 목표는 스트림이 무엇인지, 어떻게/왜 사용하는지, 사용하면 안되는 때는 없는지에 대해 알아보는 것이었습니다.
그래서 이번에는 프로젝트 내의 for 문을 최소화 하고 최대한 stream  을 사용해보려고 했는데, 이제야 적용을 해봤네요 😂

제가 스트림을 잘 몰라서인지 사용하는데 시간이 꽤 오래 걸렸던 것 같아요..!
제공하는 기능은 정말 다양하고 많은데, 어떤 기능을 지금 사용해야 하는지 잘 몰라서 많이 버벅였습니다.. ㅎㅎ

저는 사실 이렇게 모든 곳에 스트림을 남용하다 보면 스트림을 언제 사용해야하고, 언제 사용하지 말아야할지 알게 될 줄 알았습니다.
그런데 정말... 전혀 모르겠어요... ㅠㅠ

앞서 말씀드린 이펙티브 자바 스터디에서도 "스트림은 주의해서 사용해라" 라는 주제를 발표해보기도 했는데...
부끄럽게도 발표할 정도로 읽어도 언제 사용하면 안되는지 잘 모르겠습니다...

제가 아직 극단적인 예시를 만날 정도로 큰 프로젝트를 하는게 아니라서 그런걸까요?
희봉은 어떤 기준을 가지고 for 문을, 또는 stream 을 사용하시는지 궁금합니다!

<hr>

긴 글 읽어주셔서 감사합니다 :)
추가로 저번 희봉이 주신 피드백에 대한 코멘트들도 남겨두었는데, 같이 답변해주시면 감사하겠습니다!!
이번에도 날카롭고 예리한 피드백 잘 부탁드려요 ❤️

heebong commented on src/test/java/lotto/model/MoneyTest.java:
```diff
@@ -34,7 +31,23 @@ void unit_exception() {
+		assertThat(money.rate(1000)).isEqualTo(0.5);
 	}
 
 	@DisplayName("2000원치 로또가 2장인지 확인한다")
```
`money.countAvailable()` 테스트 함수 중 나누어 떨어지지 않는 경우도 있으면 좋겠네요 😄 
예를들면 `3000원` `money`에 `1100원` `price`를 넘길때..?

heebong commented on src/test/java/lotto/model/lotto/WinningBallsTest.java:
```diff
@@ -0,0 +1,55 @@
+	@Test
+	public void size_7_exception() {
+		assertThatThrownBy(() -> {
+			WinningBalls.from(new ArrayList<>(Arrays.asList("1", "2", "3", "4", "5", "6", "7")));
```
`new ArrayList<>()`로 왜 한 번 더 감쌌을까요? 👀

heebong commented on src/main/java/lotto/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,89 @@
+		return money.countAvailable(PRICE);
+	}
+
+	public int match(WinningBalls winningBalls) {
```
`Lotto`와 `WinningBalls` 모두 `match()`함수를 가지고있어요. 그리고 `Lotto.match()`가 `WinningBalls`를 받고있고요. 과연 `Lotto.match()`가 `WinningBalls`를 받는게 맞을까요? 🤔 
의존관계를 파악해봅시다. `WinningBalls`는 `Lotto`를 가지고있어요. 때문에 `WinningBalls`가 `Lotto`보다 더 상위 레벨, 혹은 더 구체적이라고 말할 수 있어요. 이렇듯 `WinningBalls`가 `Lotto`를 의존하는 상황에서, `Lotto`도 `match()` 함수로 인해 `WinningBalls`를 의존하게 되습니다. 둘은 서로를 의존하고 있네요 😂 

서로가 서로를 의존하면 왜 안좋을까요? 의존은 `변경의 전파`를 의미합니다. `Lotto` 객체에 변경이 일어나면 `Lotto`를 의존하는 모든 객체들도 그에 맞춰 변경이 일어날 수 있어요. `Lotto` 객체의 변경은 단순히 `Lotto` 하나만의 변경이 아닙니다. 그래서 서로를 의존할 수 밖에 없는 객체지향 코드는 계층을 나눠 의존관계를 최소로 만들려고 노력해요. 도메인이 변경되었지만 `dto`가 있어 `controller`와 `view`에 변화가 없었다고 느꼈던 이유도 `dto`를 통해 의존관계를 느슨하게 만들었기 때문이에요. 그래서 서로가 서로를 의존하는 코드는 피하는게 좋습니다.

heebong commented on src/main/java/lotto/model/lotto/WinningBalls.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.util.List;
+
+public class WinningBalls {
```
`WinningBalls`가 따로 필요하다고 느낀 이유는 무엇일까요? 🤔 
현재 `Lotto`와 `WinningBalls`가 하는 일은 거의 똑같습니다. `Lotto`의 `countTickets()` 함수를 제외하곤 거의 동일해요. 만약 우승번호도 자동으로 만들도록 요구사항이 들어온다면 `Lotto.fromAuto()`함수를 바이패스, 즉 전달하는 함수가 생길겁니다. 
물론 바이패스하는 함수가 나쁜건 아니에요. 중요한건 객체의 역할이에요. 지금 보기에는 `Lotto`와 `WinningBalls`의 역할이 똑같아요. 야호가 생각한 `WinningBalls`만의 역할이 있을 겁니다. 그게 어떤것일까요? 

advance ⬆️ `우승번호` 객체가 맞춘 개수를 계산하는 역할을 가지면 어떨까요? 맞춘 개수 뿐만 아니라 보너스볼의 여부도 알려주는것이죠. `우승번호` 객체가 우승번호 6개와 보너스볼 번호까지 가지고있고, `Lotto`를 받으면 맞춘 개수나 보너스 볼 여부도 알려줄 수 있을 것 같아요. 바로 `Prize`를 리턴할 수도 있겠네요 😄

heebong commented on src/main/java/lotto/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,89 @@
+		this.lotto = List.copyOf(numbers);
+	}
+
+	public static Lotto fromAuto() {
```
함수로 분리하고 `팩토리 메서드`로 잘 생성하셨네요 👍 새로운 `public` 함수로 나와있으니 테스트하기도 쉬워졌을 거에요.

advance ⬆️  클래스로 분리하라는 의미는 이런 `팩토리 메서드`를 가진 `팩토리 클래스`를 생성해보자는 의미였어요. `XXXFactory.class`라는 객체들이 존재하는데 이건 다양한 `XXX`객체를 생성하는 클래스랍니다. 지금은 `Lotto`가 생성의 책임을 가지고있지만, `팩토리 클래스`를 사용하면 생성하는 방법조차 다른 객체에게 위임할 수 있게 되는것이죠. 물론 `팩토리 클래스`가 정답이라는건 아닙니다! 클래스로 분리한다는건 이걸 의미한다는 걸 말씀드리고 싶었어요. 🙂

heebong commented on src/main/java/lotto/model/prize/PrizeInformation.java:
```diff
@@ -1,32 +1,63 @@
 public class PrizeInformation {
-	private final Prize prize;
-	private final int count;
+	private final EnumMap<Prize, Integer> prizeInformation;
```
`EnumMap` 사용 👍  일반 `Map`과 `EnumMap`은 어떤것이 다를까요?

hint 책 [이펙티브 자바]의 `아이템 36 비트 필드 대산 EnumSet을 사용하라`
`아이템 37 ordinal 인덱싱 대신 EnumMap을 사용하라`

heebong commented on src/main/java/lotto/model/dto/LottoDTO.java:
```diff
@@ -3,23 +3,24 @@
 public class LottoDTO {
 
-	private final List<Integer> numbers;
+	private final List<LottoBall> lottoBalls;
```
`Lotto`는 `DTO`가 있지만, `LottoBall`은 도메인을 그냥 넘기네요 👀

heebong replied:
`LottoBallDto`를 새로 생성하셨네요 👍 하지만 여전히 `LottoDTO`는 `LottoBall`도메인을 가지고 있어요.
DTO을 사용하는 이유는 무엇일까요? 야호가 도메인 객체를 변경했지만 DTO가 있어서 컨트롤러와 뷰는 변경이 없어 편리했다고 하셨죠? 이렇듯 DTO는 각 계층을 분리해서 변경을 막아주는 역할을 합니다. 그 역할을 기대하며 DTO를 사용하고요. 
`Lotto`는 DTO를 사용해 잘 분리했어요. `Lotto` 객체가 변경되어도 view는 변경이 없을거에요. 하지만 `LottoBall`은 어떨까요?

물론 지금은 야호가 코드를 잘 짜서 view에서 `LottoBall`객체를 직접적으로 사용하진 않아요. 하지만 언제든 view가 `LottoBall`을 직접 사용할 수 있습니다. (나중에 다른사람이, 혹은 야호가 이 사실을 까먹고 `LottoBall`을 직접 사용할 수 있어요.) 그러면 애써 DTO를 만든 이유가 사라집니다. `LottoBall`이 변경되면 view도 변경될 수 있어요.

DTO를 만든 목적을 달성하기 위해서는 어떻게 개발해야할까요? 바로 정답을 말씀드리면 `LottoDTO`가 `LottoBall`이 아닌 `LottoBallDTO`를 필드로 가지면 됩니다. 그러면 `LottoDto`를 사용하는 view에서는 LottoBall을 전혀 알 수 없겠죠. 도메인과 뷰의  의존을 dto로 완전히 없앨 수 있습니다.

바로 정답을 말씀드리긴 했지만, 야호가 이해했으면 좋겠어요. 왜 `LottoDTO`가 `LottoBall`을 가지고 있으면 안되는지. 어떤 문제가 생길지. 실은 DTO의 필요성을 느끼지 못한다면 사용하지 않는 것을 추천합니다. 하지만 야호가 DTO 사용의 장점을 느꼈기 때문에 말씀드려요.  왜 dto가 필요한지, 그렇다면 어떻게 개발해야 하는지 생각해보셨으면 좋겠습니다. 😄

heebong commented on src/main/java/lotto/model/lotto/Lotto.java:
```diff
@@ -0,0 +1,89 @@
+	public int match(WinningBalls winningBalls) {
+		return (int)lotto.stream()
+			.filter(winningBalls::match)
+			.count();
```
최대한 `stream()`으로 만든 것 👍 
저는 대부분의 경우 `for문` 말고 `stream()`을 사용합니다. 😂 인덱스 값이 필요한 경우를 제외하면 거의 `stream()`으로 만들어요.
제가 생각하기에 `stream()`의 장점은 `for문보다 코드를 줄여`주는 거라고 생각해요. 코드가 줄어드니 더 빠르고 편리하게 개발할 수 있죠. 그리고 `filter()`, `map()`같은 기능이 있어 가독성도 좋고요. 위의 코드를 보면 간단하게 `리스트 변수 lotto 중에 winning.match()가 true인 요소의 개수를 구하는 로직`이라고 알 수 있잖아요? 그래서 저는 `stream`을 많이 사용해요

하지만 이게 단점이 될 수도 있어요. 상황에 따라서는 `stream()`을 사용하는게 `for문`보다 더 길수도 있어요. 그리고 로직이 복잡해질 수록 `stream()`은 가독성이 떨어진답니다. 아무리 한 줄로 코딩할 수 있어도 가독성이 떨어진다면 `for문`으로 개발해요. 

결론은 가독성에 맡겨라 입니다. 하지만 대부분의 경우 `stream()`이 더 가독성이 좋기 때문에 `stream()`으로 개발하되, 무조건 `stream()`을 써야하는게 아니라는 점을 기억하며 개발하면 좋을 것 같아요 😄 

자바는 스트림, 람다 등이 나오면서 완전히 바뀌었어요. 자바8 이후로 정말 편해졌고 좋은 API들이 많이 생겼습니다. 따라서 스트림 등 자바8 이후에 나온 개념들을 익히는게 중요합니다. 물론 한 번에 알긴 어려워요 😢 그래서 책 [모던 자바 인 액션]을 추천합니다. 읽어보면 스트림, 람다 등 개념을 알아가는데 도움이 될거에요. 지금 이펙티브 자바 읽기처럼 크루들과 스터디로 같이 읽어도 좋을 것 같아요 😄

heebong reviewed (changes requested):
안녕하세요, 야호! 리뷰어 희봉입니다. 😄 
피드백을 잘 반영하셨네요 👍 피드백 뿐 아니라 더 필요하다고 느낀 부분도 함께 리팩토링 한 것 💯 코드가 더욱 깔끔해졌습니다. (그 과정에서 의존관계에 대해 깨달은 것도 👍)
몇가지 코멘트를 추가로 남겼습니다. `stream()` 관련 질문주신 것도 코멘드로 남겼습니다. 확인 부탁드려요.
언제든 궁금한게 있으면 댓글 혹은 DM 보내주세요! ✉️

yaho99 committed:
refactor: rate 에서 소수점 3자리까지만 반환하도록 수정, 테스트 진행

yaho99 committed:
refactor: WinningBalls 와 BonusBall 병합

1. WinningBalls 에 MatchResult 역할 포함
2. 보너스 번호와 당첨 번호에 대한 match 역할 Lotto 로 이동
3. 당첨 등수를 알려주는 역할 WinningBalls 로 이동

yaho99 committed:
refactor: LottoBallDTO 생성, LottoBall 에서 바로 값을 꺼내는 부분 수정

yaho99 committed:
refactor: map의 업데이트 값을 새로 put 하지 않고 merge 사용

yaho99 committed:
docs: 희봉의 세번째 리뷰에 따른 수정사항 체크

yaho99 commented:
안녕하세요 희봉! 
마지막 PR 이라 아쉽네요...ㅜㅜ
그동안 좋은 리뷰와 생각할 거리들 알려주셔서 감사합니다 :) ❤️

희봉이 알려주신 수정 사항 적용해 봤습니다! 
다만 제가 DTO 를 직접 짜본 것이 처음이라 제대로 만든건지 잘 모르겠어요 :(
희봉이 보시고 잘 사용한건지 알려주시면 감사하겠습니다!!

heebong commented on src/main/java/lotto/model/lotto/WinningBalls.java:
```diff
@@ -0,0 +1,44 @@
+		}
+	}
+
+	public Prize getPrize(Lotto lotto) {
```
prize를 확인하는 로직이 훨씬 간단해졌네요 👍

heebong commented on src/main/java/lotto/model/lotto/LottoBall.java:
```diff
@@ -0,0 +1,62 @@
+
+import java.util.Objects;
+
+public class LottoBall implements Comparable<LottoBall> {
```
advance ⬆️ `LottoBall`은 `equals()`, `hashcode()` 함수를 재정의 했네요 👍 
하지만 다른 객체들은 대부분 재정의 하지 않았고, `toString()`도 없습니다.
책 [이펙티브 자바]의 아래 아이템들을 읽어볼까요?
`아이템 10 equals는 일반 규약을 지켜 재정의하라`
`아이템 11 equals를 재정의하려거든 hashCode도 재정의하라`
`아이템 12 toString은 항상 재정의하라`

heebong reviewed (approved):
안녕하세요, 야호! 리뷰어 희봉입니다. 😄 
로또 미션 고생하셨습니다 🎉 단계별로 요구사항이 달라 코드를 많이 변경해야 했을 텐데 잘 작업해주셨어요.
피드백도 잘 반영하시고 추가적으로 리팩토링도 진행해서 코드가 많이 깔끔해졌습니다 👍 
이번 미션은 여기서 머지하도록 할게요. dto 사용 등 추가로 생각하면 좋을 내용을 코멘트로 짧게 남겼습니다. 다음 미션을 진행하면서 생각해보시면 좋을 것 같아요. 🙂 
고생하셨습니다 💸
